cl/persistence/beacon_indexes/sql_indexer.go

```diff
@@ -0,0 +1,138 @@
+`
+
+type SqliteBeaconIndexer struct {
+	db *sql.DB
+}
+
+func NewSqlBeaconIndexer(db *sql.DB) (SqliteBeaconIndexer, error) {
+	if _, err := db.Exec(beaconIndiciesCreateTableQuery); err != nil {
+		return SqliteBeaconIndexer{}, err
+	}
+	if _, err := db.Exec(beaconIndiciesUniqueIndiciesQuery); err != nil {
+		return SqliteBeaconIndexer{}, err
+	}
+	// Initialize the table for indicies
+	return SqliteBeaconIndexer{
+		db: db,
+	}, nil
+}
+
+func (s SqliteBeaconIndexer) ReadBlockSlotByBlockRoot(blockRoot libcommon.Hash) (uint64, error) {
+	var slot uint64
+
+	// Execute the query.
+	err := s.db.QueryRow(beaconReadBlockSlotByBlockRoot, blockRoot[:]).Scan(&slot) // Note: blockRoot[:] converts [32]byte to []byte
+	if err != nil {
+		if err == sql.ErrNoRows {
+			return 0, fmt.Errorf("no block found with the provided BeaconBlockRoot: %v", err)
+		}
+		return 0, fmt.Errorf("failed to retrieve slot for BeaconBlockRoot: %v", err)
+	}
+
+	return slot, nil
+}
+
+func (s SqliteBeaconIndexer) ReadCanonicalBlockRoot(slot uint64) (libcommon.Hash, error) {
+	blockRootBytes := make([]byte, 32)
+
+	// Execute the query.
+	err := s.db.QueryRow(beaconReadCanonicalBlockRootBySlot, slot).Scan(&blockRootBytes)
+	if err != nil {
+		if err == sql.ErrNoRows {
+			return libcommon.Hash{}, fmt.Errorf("no canonical block found for the provided slot: %v", err)
+		}
+		return libcommon.Hash{}, fmt.Errorf("failed to retrieve BeaconBlockRoot for slot: %v", err)
+	}
+
+	// Convert retrieved []byte to [32]byte and return
+	var blockRoot libcommon.Hash
+	copy(blockRoot[:], blockRootBytes)
+
+	return blockRoot, nil
+}
+
+func (s SqliteBeaconIndexer) MarkRootCanonical(slot uint64, blockRoot libcommon.Hash) error {
+	// First, reset the Canonical status for all other block roots with the same slot
+	if _, err := s.db.Exec(resetOtherRootsCanonicalQuery, slot, blockRoot[:]); err != nil {
+		return fmt.Errorf("failed to reset canonical status for other block roots: %v", err)
+	}
+
+	// Next, mark the given blockRoot as canonical
+	if _, err := s.db.Exec(markRootCanonicalQuery, slot, blockRoot[:]); err != nil {
+		return fmt.Errorf("failed to mark block root as canonical: %v", err)
+	}
+
+	return nil
+}
+
+func (s SqliteBeaconIndexer) GenerateBlockIndicies(block *cltypes.BeaconBlock) error {
+	blockRoot, err := block.HashSSZ()
+	if err != nil {
+		return err
+	}
+	_, err = s.db.Exec(insertOrUpdateBlockRoot, block.Slot, blockRoot[:], block.StateRoot[:], block.ParentRoot[:])
+	if err != nil {
+		return fmt.Errorf("failed to write block root to beacon_indicies: %v", err)
+	}
+
+	return nil
+}
+
+func (s SqliteBeaconIndexer) ReadParentBlockRoot(blockRoot libcommon.Hash) (libcommon.Hash, error) {
+	var parentRoot libcommon.Hash
+
+	// Execute the query.
+	err := s.db.QueryRow(beaconReadParentRootByBlockRoot, blockRoot[:]).Scan(parentRoot[:])
+	if err != nil {
+		if err == sql.ErrNoRows {
+			return libcommon.Hash{}, fmt.Errorf("no block found with the provided BeaconBlockRoot: %v", err)
+		}
```

Comment:
convention -  the check for this should be above err != nil.

also - never compare errors using == 

only use `errors.Is`